### PR TITLE
api: update terminology; problem -> exercise.

### PR DESCRIPTION
**Pull request by robphoenix**
see #252

### CODE REVIEW

**Comment by robphoenix** [comment]
Please see this as a work in progress, I wanted to get feedback that this is in the right direction before I continued through the rest of the files, hence such a sparse commit message.

Firstly, this commit also includes some changes that I have split off into #372 & #373 so at the very least these should be resolved to rebase this against.

To make the `api` tests pass I have changed occurrences of the word problem(s) to exercise(s) in the `fixtures/*.json` files. I don't know the exercism.io API at all and my assumption is that it returns json with the word problem(s) rather than exercise(s), therefore these changes will just break everything and make everyone unhappy :cry: 
If this is the case, does the API need to be updated concurrently?
If not, should this just be put on hold, as it's perhaps not preferable to then have the term exercises somewhere (wherever we can) and problems elsewhere (to satisfy the API)?
I'm also aware the website/API/CLI are due to change in the (near) future, which may be better to wait for?

**Comment by Tonkpils** [comment]
> To make the api tests pass I have changed occurrences of the word problem(s) to exercise(s) in the fixtures/*.json files. I don't know the exercism.io API at all and my assumption is that it returns json with the word problem(s) rather than exercise(s), therefore these changes will just break everything and make everyone unhappy 😢
> If this is the case, does the API need to be updated concurrently?
> If not, should this just be put on hold, as it's perhaps not preferable to then have the term exercises somewhere (wherever we can) and problems elsewhere (to satisfy the API)?
> I'm also aware the website/API/CLI are due to change in the (near) future, which may be better to wait for?

We can always leave the `json` tags reflect what the API return until the API is updated with the new terminology. This looks good though.

**Comment by robphoenix** [comment]
OK, grand, I'll update the tags and keep working on this.

**Comment by robphoenix** [comment]
Currently all tests are passing...

```
~/code/go/src/github.com/robphoenix/cli [domain-concepts/exercise ≡ +0 ~26 -0] 
> go test ./...                                                                                                                             [148]
zsh: correct './...' to './..' [nyae]? n
ok  	github.com/robphoenix/cli/api	0.040s
ok  	github.com/robphoenix/cli/cmd	0.013s
ok  	github.com/robphoenix/cli/config	0.006s
ok  	github.com/robphoenix/cli/debug	0.016s
?   	github.com/robphoenix/cli/exercism	[no test files]
ok  	github.com/robphoenix/cli/paths	0.010s
ok  	github.com/robphoenix/cli/user	0.005s
?   	github.com/robphoenix/cli/vendor/github.com/kardianos/osext	[no test files]
?   	github.com/robphoenix/cli/vendor/github.com/stretchr/testify/assert	[no test files]
?   	github.com/robphoenix/cli/vendor/github.com/urfave/cli	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/net/html	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/net/html/atom	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/net/html/charset	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding/charmap	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding/htmlindex	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding/ianaindex	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding/internal	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding/internal/identifier	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding/japanese	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding/korean	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding/simplifiedchinese	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding/traditionalchinese	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/encoding/unicode	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/internal/tag	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/internal/utf8internal	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/language	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/language/display	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/runes	[no test files]
?   	github.com/robphoenix/cli/vendor/golang.org/x/text/transform	[no test files]
```
I'm going to leave this for the weekend and then come back to it to double check what I've done with fresh eyes. I also haven't yet actually used it, as such, so I want to try it out in a fresh VM.
Currently the import links currently point to my copy of `cli` rather than the exercism version, so this will need to be updated.

File: config/config_test.go [hunk 1]
```diff
@@ -107,7 +107,7 @@ func TestLoad_InvalidJSON(t *testing.T) {
 		t.Fatal(err)
 	}
 	invalidPath := filepath.Join(tmpDir, "config_invalid.json")
-	if err := os.Link(fixturePath(t, "config_invalid.json"), invalidPath); err != nil {
+	if err := os.Symlink(fixturePath(t, "config_invalid.json"), invalidPath); err != nil {
```
Comment (robphoenix) [inline config/config_test.go]: ditto

File: config/config_test.go [hunk 2]
```diff
@@ -23,11 +23,11 @@ func TestLoad(t *testing.T) {
 	paths.DefaultConfig = filepath.Join(tmpDir, "default.json")
 
 	configPath := filepath.Join(paths.ConfigHome, "config.json")
-	if err := os.Link(fixturePath(t, "config.json"), configPath); err != nil {
+	if err := os.Symlink(fixturePath(t, "config.json"), configPath); err != nil {
 		t.Fatal(err)
 	}
 	dirtyPath := filepath.Join(paths.ConfigHome, "dirty.json")
-	if err := os.Link(fixturePath(t, "dirty.json"), dirtyPath); err != nil {
+	if err := os.Symlink(fixturePath(t, "dirty.json"), dirtyPath); err != nil {
```
Comment (robphoenix) [inline config/config_test.go]: ditto

**Comment by robphoenix** [comment]
:skull_and_crossbones:  DON'T TRY OUT THIS PR https://github.com/exercism/exercism.io/issues/3388

**Comment by robphoenix** [comment]
I'm going to close this and open a new PR it when I've got something more robust to submit.

**Pull request closed without merging**